storage/innobase/include/trx0trx.h

```diff
@@ -902,6 +903,146 @@ struct TrxVersion {
 
 typedef std::list<TrxVersion, ut_allocator<TrxVersion> > hit_list_t;
 
+/** Utility class to collect and report slow query log InnoDB extension
+stats */
+class trx_stats {
+	public:
+	trx_stats() {
+	/* Always created in a zeroed memory block */
+#ifdef UNIV_DEBUG
+		ut_ad(lock_que_wait_ustarted == 0);
+		ut_ad(take_stats == false);
+#endif
+	}
+
+	/**
+	Register, if needed, start of an I/O read or wait.
+
+	@param	trx	transaction to account I/O to or NULL
+	@param	bytes	number of bytes to read or 0 if this is a wait for an
+	 already posted read in progress
+	@return value to be passed to end_io_read
+	*/
+	MY_NODISCARD
+	static ib_uint64_t
+	start_io_read(trx_t *trx, ulint bytes);
+
+	/**
+	Register start of an I/O read or wait.
+
+	@param	trx	transaction to account I/O to
+	@param	bytes	number of bytes to read or 0 if this is a wait for an
+	 already posted read in progress
+	@return value to be passed to end_io_read
+	*/
+	MY_NODISCARD
+	static ib_uint64_t
+	start_io_read(const trx_t &trx, ulint bytes);
+
+	/**
+	Register, if needed, end of an I/O read or wait.
+
+	@param	trx		transaction to account I/O to or NULL
+	@param	start_time	return value of start_io_read
+	*/
+	static void
+	end_io_read(trx_t *trx, ib_uint64_t start_time);
+
+	/**
+	Register end of an I/O read or wait.
+
+	@param	trx		transaction to account I/O to
+	@param	start_time	return value of start_io_read*/
+	static void
+	end_io_read(const trx_t &trx, ib_uint64_t start_time);
+
+	/**
+	Register, if needed, a single untimed I/O read.
+
+	@param	trx	transaction to account I/O to or NULL
+	@param	bytes	number of bytes read
+	*/
+	static void
+	bump_io_read(trx_t *trx, ulint bytes);
+
+	/**
+	Register a single untimed I/O read.
+
+	@param	trx	transaction to account I/O to
+	@param	bytes	number of bytes read
+	*/
+	static void
+	bump_io_read(const trx_t &trx, ulint bytes);
+
+	/**
+	Register, if needed, a start of lock wait */
+	void
+	start_lock_wait() {
+		if (UNIV_LIKELY(!take_stats))
+			return;
+		ut_ad(lock_que_wait_ustarted == 0);
+		ulint sec, ms;
+		ut_usectime(&sec, &ms);
+		lock_que_wait_ustarted =
+		    static_cast<ib_uint64_t>(sec * 1000000 + ms);
+	}
+
+	/**
+	Register, if needed, a finished lock wait.
+
+	@param	trx	transaction to account to, if needed */
+#ifdef UNIV_DEBUG
```

Comment:
I would declare `stop_lock_wait` non-const regardless of `UNIV_DEBUG` just for clarity/ symmetry reasons unless you are fighting with modern compiler warnings.

Reply:
Making non-const